ql/src/java/org/apache/hadoop/hive/ql/reexec/ReExecutionRetryLockPlugin.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.reexec;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.Driver;
+import org.apache.hadoop.hive.ql.plan.mapper.PlanMapper;
+import org.apache.hadoop.hive.ql.processors.CommandProcessorException;
+
+public class ReExecutionRetryLockPlugin implements IReExecutionPlugin {
+
+  private Driver coreDriver;
+  private int maxRetryLockExecutions = 1;
+
+  @Override
+  public void initialize(Driver driver) {
+    coreDriver = driver;
+    maxRetryLockExecutions = 1 + coreDriver.getConf().getIntVar(HiveConf.ConfVars.HIVE_QUERY_MAX_REEXECUTION_RETRYLOCK_COUNT);
+  }
+
+  @Override
+  public void beforeExecute(int executionIndex, boolean explainReOptimization) {
+  }
+
+  @Override
+  public boolean shouldReExecute(int executionNum, CommandProcessorException ex) {
+    return executionNum < maxRetryLockExecutions && ex != null &&
+        ex.getMessage().contains(Driver.SNAPSHOT_WAS_OUTDATED_WHEN_LOCKS_WERE_ACQUIRED);
+  }
+
+  @Override
+  public void prepareToReExecute() {
+  }
+
+  @Override
+  public boolean shouldReExecute(int executionNum, PlanMapper oldPlanMapper, PlanMapper newPlanMapper) {
+    return executionNum < maxRetryLockExecutions;
```

Comment:
should we re-execute when here?

Reply:
The ReExecutePluginInterface is not really straightforward. This method is called after the query is recompiled from the reExecDriver shouldReExecuteAfterCompile . Basicly it asks after the recompilation do you want to continue the execution. It is only used by the reoptimize plugin, which tries to recompile the query with different statistics (if I understand correctly) and only reexecutes if the plan did change.

Reply:
if it's only used by reoptimize plugin shouldn't we throw UnsupportedOperationException?

Reply:
No, you must return true, saying you do not care with the result of the recompile, you want the query to reexecute.

Reply:
ok :)